File: sdks/kotlin/generate.py
#!/usr/bin/env python
# /// script
# requires-python = ">=3.11"
# dependencies = ["pyyaml"]
# ///
"""Regenerate the Kotlin models from the acceleration OpenAPI spec.

The spec at ``acceleration/api/openapi.yaml`` is the source of truth. Only the models are
generated, into ``core/src/main/kotlin/.../generated``, and they are ``internal``: the request
layer is a screenful of Ktor and is hand-written, because a generated client would decide how a
token is attached and which engine to use, which is the part a phone needs to own. The output is
committed so building needs no generator.

Only what a phone may call is generated. The spec is pruned to the operations below and the
schemas they reach before the generator sees it, and an operation not marked
``x-client-accessible`` in the spec fails the run: a client SDK with a model for a server-side
operation is one step from a method that only ever answers 403. ``--check`` verifies the filter
still agrees with the spec without regenerating.

The generator is OpenAPI Generator in Docker, pinned below, so nothing is installed on the host.

Usage:
    uv run sdks/kotlin/generate.py
    uv run sdks/kotlin/generate.py --check
"""

import shutil
import subprocess
import sys
from pathlib import Path

import yaml

KOTLIN = Path(__file__).parent
ROOT = KOTLIN.parents[1]
SPEC = ROOT / "acceleration" / "api" / "openapi.yaml"
PACKAGE = "io.getstream.visionagents.core.generated"
GENERATED = KOTLIN / "core" / "src" / "main" / "kotlin" / Path(*PACKAGE.split("."))
IMAGE = "openapitools/openapi-generator-cli:v7.25.0"

# What a client is allowed to reach: opening a conversation, finding, reading and ending one,
# going back in it or branching off it, becoming a guest, and looking something up.
OPERATIONS = [
    "closeSession",
    "createGuestUser",
    "createResponse",
    "createSession",
    "forkSession",
    "getSession",
    "listResponseItems",
    "listResponses",
    "listSessions",
    "rewindSession",
    "search",
    "searchSessions",
]

# Hand-written, because OpenAPI stops at the upgrade. Audited all the same.
SOCKETS = ["watchSession"]

GENERATOR_OPTIONS = {
    "serializationLibrary": "kotlinx_serialization",
    "dateLibrary": "string",
    "enumUnknownDefaultCase": "true",
    "nonPublicApi": "true",
    "modelMutable": "false",
    "sourceFolder": "src/main/kotlin",
    "packageName": PACKAGE,
    "modelPackage": PACKAGE,
}


def operations(spec: dict) -> dict[str, dict]:
    found: dict[str, dict] = {}
    for methods in spec["paths"].values():
        for operation in methods.values():
            if isinstance(operation, dict) and "operationId" in operation:
                found[operation["operationId"]] = operation
    return found


def audit(spec: dict) -> list[str]:
    """Report the named operations the spec does not open to a client, or does not have."""
    found = operations(spec)
    complaints = []
    for name in OPERATIONS + SOCKETS:
        operation = found.get(name)
        if operation is None:
            complaints.append(f"{name} is not in the spec")
        elif not operation.get("x-client-accessible"):
            complaints.append(f"{name} is server-side only and cannot be in a client SDK")
    return complaints


def refs(node: object) -> set[str]:
    if isinstance(node, dict):
        found = {node["$ref"]} if isinstance(node.get("$ref"), str) else set()
        for value in node.values():
            found |= refs(value)
        return found
    if isinstance(node, list):
        found = set()
        for value in node:
            found |= refs(value)
        return found
    return set()


def undefault(node: object) -> object:
    """The schema with no property defaults.

    A default is what the router does when a field is left out, which is the router's to
    apply. Generated into the model it would be the value a field starts as, and a caller's
    ``False`` equal to it would be dropped from the request while the config it was meant to
    override kept its ``True``.
    """
    if isinstance(node, dict):
        return {key: undefault(value) for key, value in node.items() if key != "default"}
    if isinstance(node, list):
        return [undefault(value) for value in node]
    return node


def prune(spec: dict) -> dict:
    """The spec with only the allowed operations and what they reach."""
    paths: dict[str, dict] = {}
    for path, methods in spec["paths"].items():
        kept = {
            method: operation
            for method, operation in methods.items()
            if isinstance(operation, dict) and operation.get("operationId") in OPERATIONS
        }
        if kept:
            paths[path] = kept

    components = spec["components"]
    keep: dict[str, set[str]] = {}
    pending = refs(paths)
    while pending:
        ref = pending.pop()
        _, _, section, name = ref.split("/")
        if name in keep.setdefault(section, set()):
            continue
        keep[section].add(name)
        pending |= refs(components[section][name])

    return {
        "openapi": spec["openapi"],
        "info": spec["info"],
        "paths": paths,
        "components": {
            section: {name: undefault(components[section][name]) for name in sorted(names)}
            for section, names in keep.items()
        },
    }


def main(argv: list[str]) -> int:
    spec = yaml.safe_load(SPEC.read_text())
    complaints = audit(spec)
    if complaints:
        for complaint in complaints:
            print(complaint, file=sys.stderr)
        return 1

    if "--check" in argv:
        print(f"{len(OPERATIONS)} client operations and a socket, none server-side only")
        return 0

    work = KOTLIN / ".codegen"
    shutil.rmtree(work, ignore_errors=True)
    (work / "out").mkdir(parents=True)
    (work / "spec.yaml").write_text(yaml.safe_dump(prune(spec), sort_keys=False))

    options = ",".join(f"{key}={value}" for key, value in GENERATOR_OPTIONS.items())
    result = subprocess.run(
        [
            "docker", "run", "--rm",
            "-v", f"{work}:/work",
            IMAGE, "generate",
            "-i", "/work/spec.yaml",
            "-g", "kotlin",
            "-o", "/work/out",
            "--global-property", "models,modelDocs=false,modelTests=false",
            "--additional-properties", options,
            # An open-ended object stays JSON rather than becoming Map<String, Any>, which
            # kotlinx.serialization cannot read.
            "--type-mappings", "AnyType=JsonElement",
            "--import-mappings", "JsonElement=kotlinx.serialization.json.JsonElement",
        ]
    )
    if result.returncode != 0:
        shutil.rmtree(work, ignore_errors=True)
        return result.returncode

    produced = work / "out" / "src" / "main" / "kotlin" / Path(*PACKAGE.split("."))
    shutil.rmtree(GENERATED, ignore_errors=True)
    GENERATED.mkdir(parents=True)
    for generated in sorted(produced.glob("*.kt")):
        shutil.move(str(generated), str(GENERATED / generated.name))
    shutil.rmtree(work, ignore_errors=True)

    print(f"regenerated {GENERATED.relative_to(ROOT)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: sdks/dart/core/tool/generate.dart
// Regenerates lib/src/generated/api.dart from the acceleration OpenAPI spec.
//
// The spec at acceleration/api/openapi.yaml is the source of truth. The output is committed,
// so depending on the package needs no code generation and no build_runner.
//
// Only the operations a device is allowed to call are generated. The filter is the point: an
// operation the spec does not mark x-client-accessible is server-side only, and a client SDK
// with a method for one would only be offering its callers a 403. --check verifies the filter
// still agrees with the spec and that the committed output is what the spec generates.
//
// The session events socket is audited but not generated: OpenAPI stops at the upgrade, so it
// is hand-written in lib/src/socket.dart.
//
//     dart run tool/generate.dart
//     dart run tool/generate.dart --check
import 'dart:io';

import 'package:yaml/yaml.dart';

// What a device may reach: opening, finding, reading back, rewinding, forking and ending a
// conversation, a guest to hold one as, and looking something up. Everything else about an
// agent is its backend's to ask for and hand down.
const operations = [
  'closeSession',
  'createGuestUser',
  'createResponse',
  'createSession',
  'forkSession',
  'getSession',
  'listResponseItems',
  'listResponses',
  'listSessions',
  'rewindSession',
  'search',
  'searchSessions',
];

// Client accessible, and hand-written because they are sockets.
const sockets = ['watchSession'];

final root = File.fromUri(Platform.script).parent.parent;
final spec = File('${root.path}/../../../acceleration/api/openapi.yaml');
final target = File('${root.path}/lib/src/generated/api.dart');

void main(List<String> args) {
  if (!spec.existsSync()) {
    stderr.writeln('no spec at ${spec.path}');
    exit(1);
  }
  final document = loadYaml(spec.readAsStringSync()) as YamlMap;

  final complaints = audit(document);
  if (complaints.isNotEmpty) {
    complaints.forEach(stderr.writeln);
    exit(1);
  }

  final generated = format(Generator(document).emit());
  if (args.contains('--check')) {
    if (!target.existsSync() || target.readAsStringSync() != generated) {
      stderr.writeln('${target.path} is stale: run dart run tool/generate.dart');
      exit(1);
    }
    print(
      '${operations.length} client operations and ${sockets.length} socket, '
      'none server-side only, output up to date',
    );
    return;
  }

  target.parent.createSync(recursive: true);
  target.writeAsStringSync(generated);
  print('regenerated lib/src/generated/api.dart');
}

/// The named operations the spec does not open to a client, or does not have.
List<String> audit(YamlMap document) {
  final found = <String, YamlMap>{};
  for (final item in (document['paths'] as YamlMap).values) {
    for (final operation in (item as YamlMap).values) {
      if (operation is YamlMap && operation['operationId'] is String) {
        found[operation['operationId'] as String] = operation;
      }
    }
  }
  return [
    for (final name in [...operations, ...sockets])
      if (found[name] == null)
        '$name is not in the spec'
      else if (found[name]!['x-client-accessible'] != true)
        '$name is server-side only and cannot be in a client SDK',
  ];
}

/// Runs the output through dart format, so a regenerated file and a committed one compare
/// byte for byte.
String format(String source) {
  final scratch = Directory.systemTemp.createTempSync('vision_agents_generate');
  try {
    final file = File('${scratch.path}/api.dart')..writeAsStringSync(source);
    final result = Process.runSync('dart', [
      'format',
      '--language-version=3.10',
      '--page-width=100',
      file.path,
    ]);
    if (result.exitCode != 0) {
      stderr.writeln(result.stderr);
      exit(1);
    }
    return file.readAsStringSync();
  } finally {
    scratch.deleteSync(recursive: true);
  }
}

class Operation {
  Operation(this.id, this.method, this.path, this.definition);

  final String id;
  final String method;
  final String path;
  final YamlMap definition;
}

/// How one schema is held in Dart, and how a value of it crosses the wire.
class Kind {
  Kind(this.type, this.decode, this.encode);

  final String type;

  /// Turns a JSON expression into a Dart value, given where it was read for the error.
  final String Function(String value, String at) decode;

  /// Turns a Dart value back into JSON.
  final String Function(String value) encode;
}

class Generator {
  Generator(this.document)
    : schemas = (document['components'] as YamlMap)['schemas'] as YamlMap,
      parameters = (document['components'] as YamlMap)['parameters'] as YamlMap;

  final YamlMap document;
  final YamlMap schemas;
  final YamlMap parameters;
  final Set<String> classes = {};

  String emit() {
    final found = <Operation>[];
    for (final MapEntry(key: path, value: item) in (document['paths'] as YamlMap).entries) {
      for (final MapEntry(key: method, value: operation) in (item as YamlMap).entries) {
        if (operation is YamlMap && operations.contains(operation['operationId'])) {
          found.add(
            Operation(
              operation['operationId'] as String,
              method as String,
              path as String,
              operation,
            ),
          );
        }
      }
    }
    found.sort((a, b) => a.id.compareTo(b.id));

    final methods = [for (final operation in found) method(operation)];
    final written = <String, String>{};
    while (classes.difference(written.keys.toSet()).isNotEmpty) {
      final next = (classes.difference(written.keys.toSet()).toList()..sort()).first;
      written[next] = model(next);
    }
    final models = [for (final name in written.keys.toList()..sort()) written[name]!];

    final info = document['info'] as YamlMap;
    return '''
// Generated by tool/generate.dart from acceleration/api/openapi.yaml (${info['title']} ${info['version']}).
// Do not edit: change the spec and regenerate.
//
// Internal to the package. Every type crossing the public API is hand-written in models.dart,
// so the spec can churn without breaking a caller.

import '../errors.dart';
import '../wire.dart';

/// The client operations, one method each.
final class Operations {
  Operations(this._wire);

  final Wire _wire;

${methods.join('\n')}
}

${models.join('\n')}

${helpers.trim()}
''';
  }

  String method(Operation operation) {
    final arguments = <String>[];
    final query = <String>[];
    var path = operation.path;

    for (final raw in (operation.definition['parameters'] as YamlList? ?? YamlList())) {
      final parameter = resolveParameter(raw as YamlMap);
      final name = parameter['name'] as String;
      final kind = kindOf(parameter['schema'] as YamlMap);
      final dart = identifier(name);
      if (parameter['in'] == 'path') {
        arguments.add('required ${kind.type} $dart');
        path = path.replaceAll('{$name}', '\${Uri.encodeComponent($dart)}');
      } else {
        arguments.add('${kind.type}? $dart');
        query.add("'$name': ${queryValue(kind, dart)}");
      }
    }

    final body = operation.definition['requestBody'] as YamlMap?;
    if (body != null) {
      final schema = jsonSchema(body)!;
      final kind = kindOf(schema);
      arguments.add(body['required'] == true ? 'required ${kind.type} body' : '${kind.type}? body');
    }

    final success = successSchema(operation.definition);
    final returns = success == null ? 'void' : kindOf(success).type;
    final call = [
      "'${operation.method.toUpperCase()}'",
      "'$path'",
      "operation: '${operation.id}'",
      if (query.isNotEmpty) 'query: {${query.join(', ')}}',
      if (body != null) body['required'] == true ? 'body: body.toJson()' : 'body: body?.toJson()',
    ].join(', ');

    final signature = arguments.isEmpty ? '' : '{${arguments.join(', ')}}';
    final summary = operation.definition['summary'];
    final doc = summary == null ? '' : '  /// $summary.\n';
    if (success == null) {
      return '$doc  Future<void> ${operation.id}($signature) async {\n'
          '    await _wire.send($call);\n'
          '  }\n';
    }
    return '$doc  Future<$returns> ${operation.id}($signature) async {\n'
        '    final json = await _wire.send($call);\n'
        '    return ${kindOf(success).decode('json', operation.id)};\n'
        '  }\n';
  }

  String model(String name) {
    final schema = schemas[name] as YamlMap;
    final properties = schema['properties'] as YamlMap? ?? YamlMap();
    final required = {...(schema['required'] as YamlList? ?? YamlList()).cast<String>()};

    final fields = <String>[];
    final parameters = <String>[];
    final reads = <String>[];
    final writes = <String>[];
    for (final MapEntry(key: wire, value: property) in properties.entries) {
      final key = wire as String;
      final kind = kindOf(property as YamlMap);
      final dart = identifier(key);
      final at = '$name.$key';
      if (required.contains(key)) {
        fields.add('  final ${kind.type} $dart;');
        parameters.add('required this.$dart');
        reads.add("$dart: ${kind.decode("json['$key']", at)}");
        writes.add("'$key': ${kind.encode(dart)}");
      } else {
        fields.add('  final ${kind.type}? $dart;');
        parameters.add('this.$dart');
        reads.add(
          "$dart: switch (json['$key']) { null => null, final Object value => ${kind.decode('value', at)} }",
        );
        final encoded = kind.encode('value');
        writes.add(
          encoded == 'value' ? "'$key': ?$dart" : "if ($dart case final value?) '$key': $encoded",
        );
      }
    }

    final description = schema['description'];
    final doc = description is String ? '/// ${firstSentence(description)}\n' : '';
    final constructor = parameters.isEmpty
        ? '  const $name();'
        : '  const $name({${parameters.join(', ')}});';
    return '''
$doc final class $name {
$constructor

  factory $name.fromJson(Object? value) {
    final json = _object(value, '$name');
    return $name(${reads.map((read) => '$read,').join('\n')});
  }

${fields.join('\n')}

  Map<String, Object?> toJson() => {${writes.map((write) => '$write,').join('\n')}};
}
''';
  }

  Kind kindOf(YamlMap schema) {
    final ref = schema[r'$ref'] as String?;
    if (ref != null) {
      final name = ref.split('/').last;
      final target = schemas[name] as YamlMap;
      if (target['type'] == 'object' && target['properties'] != null) {
        classes.add(name);
        return Kind(name, (value, at) => '$name.fromJson($value)', (value) => '$value.toJson()');
      }
      return kindOf(target);
    }

    if (schema['oneOf'] != null) {
      return Kind('Object', (value, at) => value, (value) => value);
    }

    switch (schema['type']) {
      case 'string':
        if (schema['format'] == 'date-time') {
          return Kind(
            'DateTime',
            (value, at) => "_date($value, '$at')",
            (value) => '$value.toUtc().toIso8601String()',
          );
        }
        return Kind('String', (value, at) => "_string($value, '$at')", (value) => value);
      case 'integer':
        return Kind('int', (value, at) => "_int($value, '$at')", (value) => value);
      case 'number':
        return Kind('double', (value, at) => "_double($value, '$at')", (value) => value);
      case 'boolean':
        return Kind('bool', (value, at) => "_bool($value, '$at')", (value) => value);
      case 'array':
        final item = kindOf(schema['items'] as YamlMap);
        return Kind(
          'List<${item.type}>',
          (value, at) =>
              "[for (final item in _list($value, '$at')) ${item.decode('item', '$at[]')}]",
          (value) => item.encode('item') == 'item'
              ? value
              : '[for (final item in $value) ${item.encode('item')}]',
        );
      case 'object':
        final values = schema['additionalProperties'];
        if (values is YamlMap && values['type'] == 'string') {
          return Kind(
            'Map<String, String>',
            (value, at) => "_strings($value, '$at')",
            (value) => value,
          );
        }
        return Kind(
          'Map<String, Object?>',
          (value, at) => "_object($value, '$at')",
          (value) => value,
        );
    }
    throw StateError('no Dart type for $schema');
  }

  String queryValue(Kind kind, String value) => switch (kind.type) {
    'String' => value,
    'DateTime' => '$value?.toUtc().toIso8601String()',
    _ => '$value?.toString()',
  };

  YamlMap resolveParameter(YamlMap parameter) {
    final ref = parameter[r'$ref'] as String?;
    return ref == null ? parameter : parameters[ref.split('/').last] as YamlMap;
  }

  YamlMap? jsonSchema(YamlMap withContent) {
    final content = withContent['content'] as YamlMap?;
    final json = content?['application/json'] as YamlMap?;
    return json?['schema'] as YamlMap?;
  }

  YamlMap? successSchema(YamlMap operation) {
    for (final MapEntry(key: status, value: response)
        in (operation['responses'] as YamlMap).entries) {
      if ('$status'.startsWith('2')) {
        return response is YamlMap ? jsonSchema(response) : null;
      }
    }
    return null;
  }
}

const reserved = {'default', 'in', 'is', 'new', 'null', 'switch', 'class', 'enum', 'var', 'final'};

String identifier(String wire) {
  final parts = wire.split('_');
  final camel =
      parts.first +
      parts
          .skip(1)
          .map((part) => part.isEmpty ? '' : part[0].toUpperCase() + part.substring(1))
          .join();
  return reserved.contains(camel) ? '${camel}Value' : camel;
}

String firstSentence(String text) {
  final flat = text.replaceAll(RegExp(r'\s+'), ' ').trim();
  final end = flat.indexOf('. ');
  return end == -1 ? flat : flat.substring(0, end + 1);
}

// Decoding helpers, which say which field was unreadable rather than throwing a TypeError
// from somewhere inside a constructor.
const helpers = r'''
String _string(Object? value, String at) =>
    value is String ? value : throw UnreadableException('$at is not a string');

int _int(Object? value, String at) =>
    value is num ? value.toInt() : throw UnreadableException('$at is not a number');

double _double(Object? value, String at) =>
    value is num ? value.toDouble() : throw UnreadableException('$at is not a number');

bool _bool(Object? value, String at) =>
    value is bool ? value : throw UnreadableException('$at is not a boolean');

DateTime _date(Object? value, String at) =>
    parseRouterDate(_string(value, at)) ?? (throw UnreadableException('$at is not a timestamp'));

List<Object?> _list(Object? value, String at) =>
    value is List ? value.cast<Object?>() : throw UnreadableException('$at is not a list');

Map<String, Object?> _object(Object? value, String at) =>
    value is Map ? value.cast<String, Object?>() : throw UnreadableException('$at is not an object');

Map<String, String> _strings(Object? value, String at) => {
  for (final MapEntry(:key, :value) in _object(value, at).entries) key: _string(value, '$at.$key'),
};
''';
